File: ssn/backend-dotnet/Ssn/Controllers/SpaController.cs
using Ssn.Models;
using Ssn.ViewModels;
using Ssn.ViewModels.Api;
using System.Collections.Generic;
using System.Web.Mvc;

namespace Ssn.Controllers
{
    public class SpaController : Controller
    {
        private SsnSecurityService ssnSecurityService = new SsnSecurityService();

        private PostsManager postsManager = new PostsManager();

        private SpaInitialDataViewModel LoadData()
        {
            string loggedUser = User.Identity.Name;

            List<string> friends = ssnSecurityService.GetUsers();
            friends.Remove(User.Identity.Name);

            List<PostDto> postsDto = postsManager.GetPosts(loggedUser);
            List<Post> posts = postsDto.ConvertAll(post => new Post(post));
            return new SpaInitialDataViewModel(loggedUser, friends, posts);
        }

        [Authorize(Roles = Global.ROLE_USER)]
        public ActionResult React()
        {
            SpaInitialDataViewModel viewModel = LoadData();
            return View("~/Views/React/Index.cshtml", viewModel);
        }

        [Authorize(Roles = Global.ROLE_USER)]
        public ActionResult Angular()
        {
            SpaInitialDataViewModel viewModel = LoadData();
            return View("~/Views/Angular/Index.cshtml", viewModel);
        }
    }
}

File: ssn/backend-spring5/src/main/java/com/giapyland/ssn/presentation/SpaController.java
package com.giapyland.ssn.presentation;

import java.security.Principal;
import java.util.ArrayList;
import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.giapyland.ssn.SsnApplication;
import com.giapyland.ssn.business.PostsManager;
import com.giapyland.ssn.business.SsnSecurityService;
import com.giapyland.ssn.business.model.PostDto;
import com.giapyland.ssn.presentation.api.model.Post;
import com.giapyland.ssn.presentation.model.SpaInitialData;

@Controller
public class SpaController {

	@Autowired
	private SsnSecurityService ssnSecurityService;
	@Autowired
	private PostsManager postsManager;

	private SpaInitialData loadData(String loggedUser) {
		List<String> friends = ssnSecurityService.getUsers();
		friends.remove(loggedUser);

		List<PostDto> postDtos = postsManager.getPosts(loggedUser);
		List<Post> posts = new ArrayList<Post>();
		for (PostDto postDto : postDtos)
			posts.add(new Post(postDto));

		try {
			return new SpaInitialData(loggedUser, friends, posts);
		} catch (JsonProcessingException ex) {
			SsnApplication.LOGGER.error(ex.getMessage());
			return null;
		}
	}

	@RequestMapping(value = { "/react", "/react/index", "/react/index.html" })
	@GetMapping
	public String react(Model model, Principal principal) {
		model.addAttribute("initialData", loadData(principal.getName()));
		return "react/index";
	}

	@RequestMapping(value = { "/angular", "/angular/index", "/angular/index.html" })
	@GetMapping
	public String angular(Model model, Principal principal) {
		model.addAttribute("initialData", loadData(principal.getName()));
		return "angular/index";
	}
}
